botorch_community/acquisition/scorebo.py

```diff
@@ -0,0 +1,160 @@
+
+        super().__init__(model=model)
+        # To enable fully bayesian GP conditioning, we need to unsqueeze
+        # to get num_optima x num_gps unique GPs
+        self.maximize = maximize
+        if not self.maximize:
+            optimal_outputs = -optimal_outputs
+
+        # inputs come as num_optima_per_model x num_models x d
+        # but we want it four-dimensional to condition one per model.
+
+        self.optimal_outputs = optimal_outputs.unsqueeze(-2)
+        # JES-like version of SCoreBO if optimal inputs are provided
+        if optimal_inputs is not None:
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore")
+                with fantasize_flag():
+                    with settings.propagate_grads(False):
+                        # We must do a forward pass one before conditioning.
+                        self.model.posterior(
+                            self.model.train_inputs[0], observation_noise=False
+                        )
+
+            self.optimal_inputs = optimal_inputs.unsqueeze(-2)
+            self.conditional_model = self.model.condition_on_observations(
+                X=self.model.transform_inputs(self.optimal_inputs),
+                Y=self.optimal_outputs,
+                noise=torch.full_like(self.optimal_outputs, MIN_INFERRED_NOISE_LEVEL),
+            )
+        # otherwise, we do a MES-like variant (which places vastly more emphasis on
+        # HP learning as supposed to optimization)
+        else:
+            self.conditional_model = self.model
+
+        self.normal = torch.distributions.Normal(
+            torch.zeros(
+                1, device=self.optimal_outputs.device, dtype=self.optimal_outputs.dtype
+            ),
+            torch.ones(
+                1, device=self.optimal_outputs.device, dtype=self.optimal_outputs.dtype
+            ),
+        )
+        if distance_metric not in DISTANCE_METRICS.keys():
+            raise ValueError(
+                f"Distance metric need to be one of {list(DISTANCE_METRICS.keys())}"
+            )
+        self.distance = DISTANCE_METRICS[distance_metric]
+        self.set_X_pending(X_pending)
+
+    @concatenate_pending_points
+    @t_batch_mode_transform()
+    def forward(self, X: Tensor) -> Tensor:
+        # since we have two MC dims (over models and optima), we need to
+        # unsqueeze a second dim to accomodate the posterior pass
+        prev_posterior = self.model.posterior(
+            X.unsqueeze(MCMC_DIM), observation_noise=True
+        )
+        noiseless_posterior = self.conditional_model.posterior(
+            X.unsqueeze(MCMC_DIM), observation_noise=False
+        )
+        posterior = self.model.likelihood(noiseless_posterior.mvn)
+
+        cond_means = posterior.mean.unsqueeze(-1)
+        marg_mean = prev_posterior.mean.mean(dim=MCMC_DIM, keepdim=True)
+        cond_variances = posterior.variance.unsqueeze(-1)
+        cond_covar = posterior.covariance_matrix
```

Comment:
Sorry, what is meant by _not passing in_ `X`? The code you linked does look a little bit quirky. What specifically should not be included in the likelihood call? 

I'm not quite sure I follow along with the concern!

Reply:
I finally got around to testing this a bot. Turns out `posterior = self.model.likelihood(noiseless_posterior.mvn)` introduces some bugs if the model uses outcome transforms, since the GPyTorch modules operate in the transformed space but the posterior is returned after untransforming the MVN. I'll push some changes reverting this and fixing sphinx etc, and finally land this :)